File: src/providers/nearai.ts
import axios from "axios";
import crypto from "crypto";
import type { AttestationReport, ServiceProvider } from "../types";

export class NearaiProvider implements ServiceProvider {
	name = "nearai";
	private apiBase = "https://cloud-api.near.ai/v1";

	async fetchReport(modelId: string): Promise<AttestationReport> {
		const nonce = crypto.randomBytes(32).toString("hex");

		const params = {
			model: modelId,
			signing_algo: "ecdsa",
			nonce: nonce,
		};

		console.log(
			`[Near] Fetching report for ${modelId} with nonce ${nonce.substring(0, 8)}...`,
		);

		const response = await axios.get(`${this.apiBase}/attestation/report`, {
			params,
		});
		const data = response.data;

		// Near returns array of attestations. We pick the first one for now.
		// In strict verifiable flow, you might verify all or specific signing address.
		if (
			!data.model_attestations ||
			!Array.isArray(data.model_attestations) ||
			data.model_attestations.length === 0
		) {
			throw new Error("Near report missing model_attestations");
		}

		const firstAttestation = data.model_attestations[0];

		return {
			intel_quote: firstAttestation.intel_quote,
			nvidia_payload: firstAttestation.nvidia_payload,
			raw: data,
		};
	}

	async listModels(): Promise<string[]> {
		console.log(`[Near] Fetching models from ${this.apiBase}/model/list`);
		const response = await axios.get(`${this.apiBase}/model/list`);
		const data = response.data;

		// Map response to just model IDs
		// Assuming data is array or wrapped array
		const models = Array.isArray(data) ? data : data.models || [];
		// Extract 'id' if objects, or return strings if strings
		// biome-ignore lint/suspicious/noExplicitAny: API response is dynamic
		return models.map((m: any) => (typeof m === "string" ? m : m.modelId));
	}
}

File: confidential_verifier/providers/nearai.py
import requests
import secrets
from typing import List, Dict, Any
from .base import ServiceProvider
from ..types import AttestationReport


class NearaiProvider(ServiceProvider):
    def __init__(self):
        self.api_base = "https://cloud-api.near.ai/v1"

    def fetch_report(self, model_id: str) -> AttestationReport:
        nonce = secrets.token_hex(32)
        params = {"model": model_id, "signing_algo": "ecdsa", "nonce": nonce}

        url = f"{self.api_base}/attestation/report"
        print(f"[Near] Fetching report for {model_id} with nonce {nonce[:8]}...")

        response = requests.get(url, params=params)
        response.raise_for_status()
        data = response.json()

        attestations = data.get("model_attestations", [])
        if not attestations or not isinstance(attestations, list):
            raise Exception("Near report missing model_attestations")

        first = attestations[0]
        nvidia_payload = first.get("nvidia_payload")
        if isinstance(nvidia_payload, str):
            try:
                import json

                nvidia_payload = json.loads(nvidia_payload)
            except:
                pass

        return AttestationReport(
            provider="nearai",
            intel_quote=first["intel_quote"],
            nvidia_payload=nvidia_payload,
            raw=data,
        )

    def list_models(self) -> List[str]:
        url = f"{self.api_base}/model/list"
        print(f"[Near] Fetching models from {url}")
        response = requests.get(url)
        response.raise_for_status()
        data = response.json()

        models = data if isinstance(data, list) else data.get("models", [])
        return [m if isinstance(m, str) else m.get("modelId") for m in models]
